assessments/projects/coredns/self-assessment.md

```diff
@@ -0,0 +1,268 @@
+On startup, CoreDNS runs servers, each defined by the zones and ports it serves. Each server contains a list of external plugins and manages its own plugin chain that defines how a query is processed. Plugins are server components and not individual actors since a vulnerability in a plugin could compromise the server, vice versa.
+
+### Actions
+* Create DNS servers: To handle DNS queries and tasks for other services running in a network.
+* Process DNS queries: Procss DNS queries sent by clients such as web browsers, terminal utilities, etc. These queries can be sent over UDP, TCP, or gRPC. DNS over HTTPS (DoH) or DNS over TLS (DoT) can also be used.
+* Return DNS Replies: After a plugin chain processes a DNS query, CoreDN sends an appropriate DNS reply back to the client.
+* Log data: Query, response, and error messages can be logged.
+* Collect operational metrics: The Prometheus plugin can be used to export operational metrics.
+* Handle secrets: Secrets such as private keys and credentials for the cloud services are handled and processed by CoreDNS.
+* Plugins can provide additional functionality for DNS servers. For instance, plugins can make API calls and make calls to cloud services. Certain plugins can also receive API requests from clients via HTTP. These can send back responses outside of typical DNS data flow, eg: the health plugin can requested by a client to provide the health status of the server. 
+
+### Goals
+* Primary Functionality:
+    * To serve as a flexible and efficient DNS server, especially in containerized environments like Kubernetes.
+    * Provide robust service discovery functions in distributed systems.
+* Security Guarantees:
+    * Ensure a high level of security, leveraging Go's memory-safe properties to prevent common vulnerabilities such as buffer overflows.
+    * Implement secure DNS query options like DoT, DoH, and DoQ to protect against eavesdropping and tampering.
+* Performance and Flexibility:
+    * Deliver high performance with a plugin-based architecture, allowing for custom extensions and optimizations.
+    * Support a variety of backends (e.g., etcd, Kubernetes) for diverse deployment scenarios.
+* Integration with Cloud and Container Environments:
+    * Seamless integration with cloud-native technologies, particularly Kubernetes, for effective service discovery in dynamic environments.
+* User-Friendly Configuration:
+    * Offer a simple and intuitive configuration system (Corefile), making it accessible for users with varying levels of expertise.
+
+### Non-goals
+* Full Recursive DNS Capabilities:
+    * CoreDNS does not aim to support full recursion like traditional DNS servers (e.g., BIND), relying instead on forwarders for comprehensive domain resolution.
+* Universal DNS Server Replacement:
+    * It is not intended to be a one-size-fits-all replacement for all types of DNS servers, especially in environments not utilizing containerized or microservices architectures.
```

Comment:
I'm not sure what this is saying specifically. CoreDNS can run outside of a container. What are some types of DNS servers that you perceive CoreDNS does not intended to replace?